velox/expression/tests/ArgumentTypeFuzzer.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Random.h>
+
+#include "velox/expression/FunctionSignature.h"
+#include "velox/expression/SignatureBinder.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::test {
+
+// Bind a function signature with a concrete return type. bindings() return a
+// map from each type variable in the signature to the corresponding concrete
+// type if determined. If a type variable is not bounded by the return type, it
+// maps to nullptr.
+class ReverseSignatureBinder {
+ public:
+  ReverseSignatureBinder(const exec::FunctionSignature& signature)
+      : signature_{signature},
+        argumentTypes_{},
+        binder_{signature_, argumentTypes_} {}
+
+  bool tryBindReturnType(const TypePtr& returnType) {
+    std::vector<TypePtr> argumentTypes;
+    if (!binder_.tryBind(signature_.returnType(), returnType)) {
+      return false;
+    }
+    return true;
+  }
+
+  // Return the determined bindings. This function should be called after
+  // tryBindReturnType.
+  const std::unordered_map<std::string, TypePtr>& bindings() const {
+    return binder_.bindings();
+  }
+
+ private:
+  const exec::FunctionSignature& signature_;
+
+  // Placeholder for creating binder_;
+  std::vector<TypePtr> argumentTypes_;
+
+  exec::SignatureBinder binder_;
+};
+
+class ArgumentTypeFuzzer {
+ public:
+  ArgumentTypeFuzzer(
+      const exec::FunctionSignature& signature,
+      std::mt19937& seed)
+      : signature_{signature}, argumentTypes_{}, seed_{seed} {}
+
+  // For function signatures containing type variables, try generate a list of
+  // argument types such that the function taking these arugments returns the
+  // given type. If there are type variables unbounded by the return type,
+  // generate a random type for them seed_.
+  bool fuzzArgumentTypes(const TypePtr& returnType);
+
+  // Return the generated list of argument types.
+  const std::vector<TypePtr>& argumentTypes() const {
```

Comment:
Any particular reason fuzzArgumentTypes doesn't return these?

Reply:
It's because fuzzArgumentTypes() return a bool indicating whether the fuzzing is successful. So if we'd like it to also return the list of argument types, that list needs to be an in-out parameter. I feel a separate interface may be cleaner?

Reply:
Got it. Thanks for explaining. What would cause fuzzArgumentTypes to return 'false'?

Reply:
fuzzArgumentTypes could return false if (for example) the return type signature of the function is "bigint" while it's argument returnType is DOUBLE(), or if there is an argument type signature that cannot bind to any concrete TypePtr (e.g., something illegal like "bigggint").

Reply:
> the return type signature of the function is "bigint" while it's argument returnType is DOUBLE()

I'm not sure I understand. Would you clarify a bit more?

> if there is an argument type signature that cannot bind to any concrete TypePtr (e.g., something illegal like "bigggint").

Don't we have a check in FunctionSignature constructor to catch that? If not, should we add that?

Reply:
> > the return type signature of the function is "bigint" while it's argument returnType is DOUBLE()
> 
> I'm not sure I understand. Would you clarify a bit more?
For example, when we try to fuzz the argument types of `foo(array(K)) -> map(K, bigint)` with a required return type `MAP(VARCHAR(), DOUBLE())`, i.e., we would like to find the argument type of `foo` such that `foo` returns `MAP(VARCHAR(), DOUBLE())`. This is impossible, so we return false.  

> > if there is an argument type signature that cannot bind to any concrete TypePtr (e.g., something illegal like "bigggint").
> 
> Don't we have a check in FunctionSignature constructor to catch that? If not, should we add that?
Oh yes, we did. Basically I'm just not sure whether we could assume there is always a type kind of the same name (in upper case) as a type signature (except "any" which we'll handle separately). For example, might the type signature name be a custom type name?

Reply:
Got it. Thank you for explaining.

Reply:
Just to correct myself: custom type name is not a problem since I'll use SignatureBinder::tryResolveType(). It handles them properly.